src/validation.cpp

```diff
@@ -1281,6 +1315,130 @@ static void CheckForkWarningConditionsOnNewFork(CBlockIndex* pindexNewForkTip) E
     CheckForkWarningConditions();
 }
 
+/**
+ * Called when a header (re-)reached BLOCK_VALID_TREE.
+ *
+ * setBlockIndexHeaderCandidates is a bit more complicated than
+ * setBlockIndexCandidates as setBlockIndexCandidates can be rather lazy
+ * as everything in it is about to be connected. In our case, we may have many
+ * headers above the tip leading down different chains, for which we really
+ * only want to keep the tip of each chain.
+ *
+ * Works even if chainActive is empty!
+ *
+ * If chain_ordered_inertion, we assume that if pindex->pprev was previously a
+ * header candidate, it will be when we're called. ie we assume that there are
+ * no header candidates which are parents of us except for possibly our direct
+ * parent.
+ *
+ * no_descendants allows us to make a similar, but inverted, assumption -
+ * assuming no descendant blocks may be header candidates.
+ */
+void CChainState::MaybeAddNewHeaderCandidate(CBlockIndex* pindex, bool chain_ordered_insertion, bool no_descendants) {
+    if (!pindex->IsValid(BLOCK_VALID_TREE)) return; // We only want things that have a valid header tree
+
+    bool lower_work = chainActive.Tip() != nullptr && chainActive.Tip()->nChainWork > pindex->nChainWork;
+    if (lower_work) return; // We don't want things with less work than our current tip
+
+    bool parent_present = false;
+    if (pindex->pprev && setBlockIndexHeaderCandidates.count(pindex->pprev)) {
```

Comment:
Nit: `setBlockIndexHeaderCandidates.count(pindex->pprev) > 0` to improve readability.